src/plugins/preauth/pkinit/pkinit_srv.c

```diff
@@ -1489,3 +1567,293 @@ kdcpreauth_pkinit_initvt(krb5_context context, int maj_ver, int min_ver,
+    count = 0;
+    for (mod = modules; *mod != NULL; mod++) {
+        h = k5calloc(1, sizeof(*h), &ret);
+        if (h == NULL)
+            goto cleanup;
+
+        ret = (*mod)(context, 1, 1, (krb5_plugin_vtable)&h->vt);
+        if (ret) {
+            free(h);
+            continue;
+        }
+        h->moddata = NULL;
+        if (h->vt.init != NULL) {
+            ret = h->vt.init(context, &h->moddata);
+            if (ret) {
+                free(h);
+                continue;
+            }
+        }
+        list[count++] = h;
+        list[count] = NULL;
+    }
+    list[count] = NULL;
+
+    ret = 0;
+    *handle_out = list;
+    list = NULL;
+
+cleanup:
+    k5_plugin_free_modules(context, modules);
+    free_certauth_handles(context, list);
+    return ret;
+}
+
+static krb5_error_code
+certauth_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                   krb5_data cert, krb5_principal princ, void *opts,
+                   void *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    int valid_san, valid_eku;
+    struct certauth_req_opts *req_opts = opts;
+
+    if (authinds_out != NULL)
+        *authinds_out = NULL;
+
+    /* Decode cert into reqctx. */
+    ret = crypto_decode_der_cert(context, cert, req_opts->reqctx->cryptoctx);
+    if (ret)
+        goto err;
+
+    /* Verify the client SAN. */
+    ret = verify_client_san(context, req_opts->plgctx, req_opts->reqctx,
+                            req_opts->cb, req_opts->rock, princ, &valid_san);
+    if (ret)
+        goto err;
+
+    if (!valid_san) {
+        pkiDebug("%s: did not find an acceptable SAN in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_CLIENT_NAME_MISMATCH;
+        goto err;
+    }
+
+    /* Verify the client EKU. */
+    ret = verify_client_eku(context, req_opts->plgctx, req_opts->reqctx,
+                            &valid_eku);
+    if (ret)
+        goto err;
+
+    if (!valid_eku) {
+        pkiDebug("%s: did not find an acceptable EKU in user "
+                 "certificate\n", __FUNCTION__);
+        ret = KRB5KDC_ERR_INCONSISTENT_KEY_PURPOSE;
+    }
+
+err:
+    return ret;
+}
+
+krb5_error_code
+certauth_pkinitsrv_initvt(krb5_context context, int maj_ver, int min_ver,
+                          krb5_plugin_vtable vtable)
+{
+    krb5_certauth_vtable vt;
+
+    if (maj_ver != 1)
+        return KRB5_PLUGIN_VER_NOTSUPP;
+    vt = (krb5_certauth_vtable)vtable;
+    vt->name = "pkinitsrv";
+    vt->authorize = certauth_authorize;
+    vt->init = NULL;
+    vt->fini = NULL;
+    vt->free_ind = NULL;
+    return 0;
+}
+
+/*
+ * Do certificate auth based on match expressions in the pkinit_cert_match
+ * attribute string.  An expression should be in the same form as those used for
+ * the pkinit_cert_match configuration option.  Multiple expressions are
+ * separated by ':'.  The certificate is authorized if all expressions are
```

Comment:
Do we need support for multiple expressions?  An expression can already contain multiple rules whose results can either be and-ed or or-ed together.  If we have a use case for an and-of-ors, that functionality is probably best moved into pkinit_matching.c.